lib/sync.js

```diff
@@ -36,7 +36,25 @@ var DEBUG = true;
 // beyond that and wedge forever, so we need to track how long we are willing
 // to keep open the connection. This constant is *ADDED* to the timeout= value
 // to determine the max time we're willing to wait.
-var BUFFER_PERIOD_MS = 80 * 1000;
+var BUFFER_PERIOD_MS = 5 * 1000;
+
+// if /sync fails, we call /versions in a loop until we get a response
+// (/versions typically returns much quicker than /sync, so this allows uys to
+// recover from network outages).
+//
+// These parameters define the intervals between /versions requests, and the
+// amount of time we wait for each /versions request to return. The randomness
+// factor attempts to avoid having all clients getting in sync, causing a
+// thundering herd.
+//
+// Each request to /versions is given a timeout of between KEEPALIVE_TIMEOUT_MS
+// and KEEPALIVE_TIMEOUT_MS+KEEPALIVE_RANDOMNESS_MS. If it times out, we make
+// another request immediately. If it fails, we delay between
+// KEEPALIVE_MIN_DELAY_MS and KEEPALIVE_MAX_MS before making another request.
+//
+var KEEPALIVE_MIN_DELAY_MS = 1000;
+var KEEPALIVE_TIMEOUT_MS = 2000;
```

Comment:
I wonder how easily we'll hit the 2s timeout on GPRS... perhaps 4s + rand(2s) might be a better compromise?